mirascope/tools/base.py

```diff
@@ -0,0 +1,89 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import ClassVar, Generic, TypeVar, cast
+
+from pydantic import BaseModel, ConfigDict, create_model
+
+from mirascope.core import BaseTool
+
+
+class _ToolConfig(BaseModel, ABC):
+    """Base configuration for tools"""
+
+    @classmethod
+    @abstractmethod
+    def from_env(cls) -> _ToolConfig:
+        """Load configuration from environment variables"""
+
+    @classmethod
+    def override_defaults(cls, config: _ToolConfig) -> _ToolConfig:
+        """Create config with overridden default values"""
+        return cls.model_validate(config)
+
+
+ConfigT = TypeVar("ConfigT", bound=_ToolConfig)
+
+_ToolSchemaT = TypeVar("_ToolSchemaT")
+
+
+class ConfigurableTool(BaseTool[_ToolSchemaT], Generic[ConfigT, _ToolSchemaT], ABC):
+    """
+    Abstract base class for configurable tools.
+
+    Subclasses must define a `__prompt_usage_description__` class variable
+    and __config__ class variable with a subclass of _ToolConfig.
+    """
+
+    __config__: ClassVar[_ToolConfig]
+    __prompt_usage_description__: ClassVar[str]
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    @classmethod
+    def _config(cls) -> ConfigT:
+        """Get tool configuration"""
+        return cast(ConfigT, cls.__config__)
+
+    @classmethod
+    def from_config(cls, config: _ToolConfig) -> type[ConfigurableTool]:
+        """Create tool class with custom configuration"""
+        config = cls.__config__.override_defaults(config)
```

Comment:
hmm, seems like this is the only place we use this, I wonder if having a method for this is a bit overkill, why not just call `config = cls.model_validate(config)` here instead?